### ISSUE TITLE
Slow random variate generation in scipy.stats

### ISSUE THREAD

**Issue by chrisb83** [scipy/scipy#10106]
I did a little test to check which distributions have slow sampling methods.
Results in seconds to generate 500 rvs if that takes longer than 0.01s (i.e. one million rvs takes ~20 secs assuming a linear relation)

```
{'argus': 3.0119,
 'cosine': 2.0341,
 'gausshyper': 11.5819,
 'genexpon': 5.5011,
 'gengamma': 0.0103,
 'gennorm': 0.0102,
 'kappa4': 0.0101,
 'ksone': 3.8842,
 'rdist': 2.3227,
 'semicircular': 1.6377,
 'randint': 0.0105}
```

Some of the continuous distributions should be easy to fix since they have bounded densities on a compact support and simple methods are known in that case (standard reference: Devroye - Non-Uniform RV Generation)

Highest priority might be to fix `randint`.

stats/tests/test_continuous_basic.py lists slows distributions:

distslow = ['kappa4', 'rdist', 'gausshyper',
            'recipinvgauss', 'genexpon',
            'vonmises', 'vonmises_line', 'mielke', 'semicircular',
            'cosine', 'invweibull', 'powerlognorm', 'johnsonsu', 'kstwobign']

It is not clear what "slow" means in this context: is it rvs, ppf, cdf? Potentially not all tests related to such distributions are slow whereas `test_cont_basic` is marked as slow which runs a lot of different tests.

Code to generate the results:

```
from scipy import stats
from scipy._lib._numpy_compat import suppress_warnings
import time

sampling_time = {}
alldists = distcont + distdiscrete # these list are from stats/_distr_params.py

for dist in alldists:
    distname = dist[0]
    args = dist[1]
    
    try:
        distfn = getattr(stats, distname)
    except TypeError:
        distfn = distname
        distname = 'rv_histogram_instance'
    np.random.seed(765456)
    sn = 500
    with suppress_warnings() as sup:
        # frechet_l and frechet_r are deprecated, so all their
        # methods generate DeprecationWarnings.
        sup.filter(category=DeprecationWarning, message=".*frechet_")
        tic = time.time()
        rvs = distfn.rvs(size=sn, *args)
        sampling_time[distname] = time.time() - tic

slow_dist = {}

for k, v in sampling_time.items():
    if v > 0.01:
        slow_dist[k] = round(v, 4)
```

**Comment by ev-br** [comment]
Re what is slow: I'd guess anything which foes through the generic cdf or especially ppf is slow. From memory, rvs is slow, and test_moments can be *very* slow, since it invokes `expect`.

**Comment by WarrenWeckesser** [comment]
@chrisb83, thanks, this could be a nice improvement.

The first thing to do is to follow up on @ev-br's comment and determine whether or not each distribution has an implementation of CDF and PPF.  For those that don't, double-check that there isn't a closed form expression that could be used.  It is possible that an expression exists but was not implemented in the code.  Those that really don't have an efficient PPF would be good candidates for using the ratio of uniforms algorithm for the rvs.

I know, for example, that `cosine` has a CDF but doesn't have a PPF implementation, and there isn't a closed form expression for it.  I have some scratch work on an implementation of the PPF that would need to be implemented as a ufunc in C or Cython to be efficient.  Eventually I'll convert that into a pull request, so maybe put `cosine` lower on the priority list of distributions that will use the ratio of uniforms method.

**Comment by chrisb83** [comment]
As expected, the really slow cases occur when there is no `_ppf` and no `_rvs`. Once `_ppf` is implemented, it can still be slow (~0.01 secs for 500 rvs) but much faster compared to the other cases.
Probably in that case, `_ppf` is often a complicated special function.

```
            name cdf ppf rvs     time
0          argus   1   0   0   3.0119
1         cosine   1   0   0   2.0341
2     gausshyper   0   0   0  11.5819
3       genexpon   1   0   0   5.5011
4       gengamma   1   1   0   0.0103
5        gennorm   1   1   0   0.0102
6         kappa4   1   1   0   0.0101
7          ksone   1   1   0   3.8842
8          rdist   1   0   0   2.3227
9   semicircular   1   0   0   1.6377
10       randint   1   1   1   0.0105
```

```
has_cdf = []
has_ppf = []
has_rvs = []

for d in slow_dist.keys():
    dist_attr = getattr(stats.distributions, ''.join([d, "_gen"])).__dict__.keys()
    has_cdf.append(int("_cdf" in dist_attr))
    has_ppf.append(int("_ppf" in dist_attr))
    has_rvs.append(int("_rvs" in dist_attr))
    
        
df = pd.DataFrame(np.array([list(slow_dist.keys()), has_cdf, has_ppf, has_rvs, list(slow_dist.values())])).T
df.columns = ["name", "cdf", "ppf", "rvs", "time"]
```

**Comment by chrisb83** [comment]
`ksone` is an exception: `_ppf` is implemented but is it still very slow 

all tests run with SciPy 1.2.0

**Comment by pvanmulbregt** [comment]
What shape arguments are being used with these distributions? 
E.g. for `ksone`, the default in `scipy/stats/_distr_params.py` is `n=1000`.  Make it smaller and the rvs will be generated in less time :-)
Yes `ksone`'s `_ppf` takes cycles --- it is applying N-R root-finding to invert the CDF.  Typically only a few iterations are needed, as it starts with a decent approximation to the root.  But computing the CDF takes time. I.e. `ksone._ppf` takes time because `ksone._cdf` takes time.

**Comment by chrisb83** [comment]
The parameters are from scipy/stats/_distr_params.py

**Comment by chrisb83** [comment]
- One can use a very simple rejection method for `cosine`, 1mn samples take about 2 secs in that case
- `rdist` is sometimes called symmetric beta distribution since it can be written as `2*B - 1` where B follows a beta distribution with suitable parameters => sampling is easy

**Comment by WarrenWeckesser** [comment]
>     * One can use a very simple rejection method for `cosine`, 1mn samples take
>       about 2 secs in that case

OK, but with a fast (i.e. C or Cython) implementation of the PPF, it should be able to generate samples as fast as, say, the beta distribution.

>     * `rdist` is sometimes called symmetric beta distribution since it can be written
>       as `2*B - 1` where B follows a beta distribution with suitable parameters =>
>       sampling is easy

Nice catch.  So all we need is to implement `rdist._ppf` with `scipy.special.btdtri` and `rdist.rvs` should be nice and snappy.

**Comment by WarrenWeckesser** [comment]
FYI: I have a preliminary version of an efficient PPF for the cosine distribution.  A sample timing:

```
In [9]: %timeit s = cosine.rvs(size=1000000)                                                                          
156 ms ± 130 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

With just 1000 samples (so I can compare it to the current implementation in scipy 1.2.1 in a reasonable amount of time):
```
In [10]: %timeit s = cosine.rvs(size=1000)                                                                            
185 µs ± 620 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```
Here's the same timing in scipy 1.2.1:
```
In [3]: %timeit s = cosine.rvs(size=1000)                                       
637 ms ± 5.9 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```
(Note the change in units.)

Once I've checked that the code does not unnecessarily lose precision in a couple places, I'll create a pull request with these changes.

**Comment by chrisb83** [comment]
Cool, looks like this is faster than the rejection method I implemented (probably it could be speed up if the loop were implemented in C).

```
class cosine2_gen(rv_continuous):
    def _pdf(self, x):
        return 1.0/2/np.pi*(1+np.cos(x))

    def _rvs(self):
        # use a simple rejection algorithm: note that
        # self.pdf(x) <= c * trapz(a, b, loc=-np.pi, scale=2*np.pi).pdf(x)
        # on [-pi, pi] with c = (1 + 0.25/pi), a = 0.5-0.25/np.pi,
        # b = 0.5+0.25/np.pi since
        # we can therefore apply the rejection method, see Devroye:
        # Non-Uniform Random Variate Generation, 1986, p. 42
        a, b = 0.5 - 0.25/np.pi, 0.5 + 0.25/np.pi
        g = stats.trapz(a, b, loc=-np.pi, scale=2*np.pi)
        size1d = tuple(np.atleast_1d(np.array(self._size, dtype=int)))
        N = int(np.prod(size1d))
        x = np.zeros(N)
        simulated = 0
        # apply rejection method
        while simulated < N:
            k = N - simulated
            u = self._random_state.random_sample(size=k)
            v = g.rvs(size=k)
            accept = (u * (1 + 0.5/np.pi) * g.pdf(v) <= self._pdf(v))
            num_accept = np.sum(accept)
            if num_accept > 0:
                x[simulated:(simulated + num_accept)] = v[accept]
                simulated += num_accept

        if self._size == ():
            # return scalar if rvs() is called without size specified
            return x[0]
        return np.reshape(x, size1d)

cosine2 = cosine2_gen(a=-np.pi, b=np.pi, name='cosine')

timeit(cosine2.rvs(size=1000))
# 18.6 ms ± 223 µs per loop (mean ± std. dev. of 7 runs, 100 loops each) 

timeit(stats.cosine.rvs(size=1000)) # SciPy 1.2.1
# 3.77 s ± 23.9 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

timeit(cosine2.rvs(size=1000000))
# 1.52 s ± 163 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

(look like my computer is much slower than yours :) )

**Comment by WarrenWeckesser** [comment]
@chrisb83, do you want to tackle the easy fixes to `rdist`?  If not, I can probably get to that in the next day or so.

**Comment by chrisb83** [comment]
> do you want to tackle the easy fixes to `rdist`?

yes, i will create a PR soon

**Comment by chrisb83** [comment]
Sampling of `genexpon` is described in H.K. Ryu, “An Extension of Marshall and Olkin’s Bivariate Exponential Distribution”, Journal of the American Statistical Association, 1993. (I don't understand the details yet)

**Comment by mdhaber** [comment]
Update:
```
{'cosine': 1.1247,
 'gausshyper': 4.2776,
 'genexpon': 2.409,
 'ksone': 1.5156,
 'kstwo': 2.5434,
 'nhypergeom': 0.614}
```
And `cosine` is being addressed.

**Comment by chrisb83** [comment]
The cosine distribution can be removed from the list of slow distributions after #10170 was merged and released in 1.7.0

**Comment by WarrenWeckesser** [comment]
In https://github.com/scipy/scipy/pull/17708, I provide an implementation of `_ppf` and `_isf` for `genexpon`, which makes `genexpon.rvs()` much faster.

Timing comparison

* `main` branch (version `'1.11.0.dev0+1225.50aa706'`):

  ```
  In [2]: %timeit genexpon.rvs(0.5, 2, 1, size=1000)
  961 ms ± 22.4 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
  ```

* After gh-17708:

  ```
  In [2]: %timeit genexpon.rvs(0.5, 2, 1, size=1000)
  148 µs ± 113 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
  ```
  
  Note the change in units in the `timeit` result.
